addons/l10n_it/models/declaration_of_intent.py

```diff
@@ -0,0 +1,251 @@
+        copy=False,
+    )
+
+    invoice_ids = fields.One2many(
+        'account.move',
+        'l10n_it_declaration_of_intent_id',
+        string="Invoices / Refunds",
+        copy=False,
+        readonly=True,
+    )
+
+    _sql_constraints = [
+                        ('protocol_number_unique',
+                         'unique(protocol_number_part1, protocol_number_part2)',
+                         _("The Protocol Number of a Declaration of Intent must be unique! Please choose another one.")),
+                        ('threshold_positive',
+                         'CHECK(threshold > 0)',
+                         _("The Threshold of a Declaration of Intent must be positive.")),
+                       ]
+
+    @api.depends('protocol_number_part1', 'protocol_number_part2')
+    def _compute_display_name(self):
+        for record in self:
+            record.display_name = f"{record.protocol_number_part1}-{record.protocol_number_part2}"
+
+    @api.depends('invoice_ids', 'invoice_ids.state', 'invoice_ids.amount_total_signed', 'invoice_ids.move_type')
+    def _compute_consumed_amount(self):
+        if not self.ids:
+            self.consumed_amount = False
+            return
+        domain = [
+            ('l10n_it_declaration_of_intent_id', 'in', self.ids),
+            ('state', '=', 'posted'),
+            ('move_type', 'in', self.env['account.move'].get_invoice_types()),
+        ]
+        groups = self.env['account.move']._read_group(domain, ['l10n_it_declaration_of_intent_id'], ['amount_total_signed:sum'])
+        treated = self.env[self._name]
+        for declaration, amount_total_signed_sum in groups:
+            declaration.consumed_amount = amount_total_signed_sum
+            treated |= declaration
+        (self - treated).consumed_amount = False
+
+    def _compute_committed_amount(self):
+        # to override in l10n_it_sale
+        self.committed_amount = False
+
+    @api.depends('threshold', 'committed_amount', 'consumed_amount')
+    def _compute_remaining_amount(self):
+        for record in self:
+            record.remaining_amount = record.threshold - record.consumed_amount - record.committed_amount
+
+    def _build_threshold_warning_message(self, record, updated_remaining):
+        ''' Build the warning message that will be displayed in a yellow banner on top of `record`
+            if the `remaining_amount` of the Declaration of Intent is less than 0 when including `record`.
+            :param record:                     The record where the warning will appear (Invoice, Sales Order...).
+            :param updated_remaining (float):  The `remaining_amount` when including `record`.
+            :return (str):                     The warning message to be shown.
+        '''
+        if updated_remaining > 0:
+            return ''
+        return _('Pay attention, (including this document) the applicable threshold of your Declaration of Intent number %s of %s is exceeded by %s.',
```

Comment:
Pay attention, the threshold of your Declaration of Intent %s of %s is exceeded by %s, this document included.